File: polars/expr_list.go
package polars

import pb "github.com/isesword/polars-go/proto"

// ListNameSpace exposes list expression helpers.
type ListNameSpace struct {
	expr Expr
}

type ListSortOptions struct {
	Descending bool
	NullsLast  bool
}

type ListSampleOptions struct {
	WithReplacement bool
	Shuffle         bool
	Seed            *uint64
}

// Element references the current list element inside List().Eval(...).
func Element() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_Element{
				Element: &pb.Element{},
			},
		},
	}
}

// ConcatList horizontally concatenates expressions into a list expression.
func ConcatList(exprs ...Expr) Expr {
	protoExprs := make([]*pb.Expr, 0, len(exprs))
	for _, expr := range exprs {
		protoExprs = append(protoExprs, expr.toProto())
	}
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ConcatList{
				ConcatList: &pb.ConcatList{
					Exprs: protoExprs,
				},
			},
		},
	}
}

// List enters the list namespace for list-typed expressions.
func (e Expr) List() ListNameSpace {
	return ListNameSpace{expr: e}
}

// Eval performs element-wise computation within each list.
func (ns ListNameSpace) Eval(evaluation Expr) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListEval{
				ListEval: &pb.ListEval{
					Expr:       ns.expr.toProto(),
					Evaluation: evaluation.toProto(),
				},
			},
		},
	}
}

// Agg performs list aggregation semantics within each list.
func (ns ListNameSpace) Agg(evaluation Expr) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListAgg{
				ListAgg: &pb.ListAgg{
					Expr:       ns.expr.toProto(),
					Evaluation: evaluation.toProto(),
				},
			},
		},
	}
}

// Len returns the number of values in each list.
func (ns ListNameSpace) Len() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListLen{
				ListLen: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Sum returns the sum of each sublist.
func (ns ListNameSpace) Sum() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListSum{
				ListSum: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Mean returns the mean of each sublist.
func (ns ListNameSpace) Mean() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListMean{
				ListMean: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Median returns the median of each sublist.
func (ns ListNameSpace) Median() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListMedian{
				ListMedian: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Max returns the maximum of each sublist.
func (ns ListNameSpace) Max() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListMax{
				ListMax: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Min returns the minimum of each sublist.
func (ns ListNameSpace) Min() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListMin{
				ListMin: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Std returns the sample standard deviation of each sublist.
func (ns ListNameSpace) Std(ddof ...uint8) Expr {
	value := uint32(1)
	if len(ddof) > 0 {
		value = uint32(ddof[0])
	}
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListStd{
				ListStd: &pb.ListMoment{
					Expr: ns.expr.toProto(),
					Ddof: value,
				},
			},
		},
	}
}

// Var returns the sample variance of each sublist.
func (ns ListNameSpace) Var(ddof ...uint8) Expr {
	value := uint32(1)
	if len(ddof) > 0 {
		value = uint32(ddof[0])
	}
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListVar{
				ListVar: &pb.ListMoment{
					Expr: ns.expr.toProto(),
					Ddof: value,
				},
			},
		},
	}
}

// Contains checks whether each sublist contains the provided value.
func (ns ListNameSpace) Contains(value interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListContains{
				ListContains: &pb.ListContains{
					Expr:       ns.expr.toProto(),
					Item:       exprFromValue(value).toProto(),
					NullsEqual: false,
				},
			},
		},
	}
}

// Sort sorts each sublist using the provided options.
func (ns ListNameSpace) Sort(opts ...ListSortOptions) Expr {
	var sortOpts ListSortOptions
	if len(opts) > 0 {
		sortOpts = opts[0]
	}
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListSort{
				ListSort: &pb.ListSort{
					Expr:       ns.expr.toProto(),
					Descending: sortOpts.Descending,
					NullsLast:  sortOpts.NullsLast,
				},
			},
		},
	}
}

// Reverse reverses each sublist.
func (ns ListNameSpace) Reverse() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListReverse{
				ListReverse: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Unique keeps unique values in each sublist.
func (ns ListNameSpace) Unique() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListUnique{
				ListUnique: &pb.ListUnique{
					Expr:          ns.expr.toProto(),
					MaintainOrder: false,
				},
			},
		},
	}
}

// UniqueStable keeps unique values in each sublist while preserving order.
func (ns ListNameSpace) UniqueStable() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListUniqueStable{
				ListUniqueStable: &pb.ListUnique{
					Expr:          ns.expr.toProto(),
					MaintainOrder: true,
				},
			},
		},
	}
}

// NUnique returns the number of unique values in each sublist.
func (ns ListNameSpace) NUnique() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListNUnique{
				ListNUnique: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// ArgMin returns the index of the minimum value in each sublist.
func (ns ListNameSpace) ArgMin() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListArgMin{
				ListArgMin: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// ArgMax returns the index of the maximum value in each sublist.
func (ns ListNameSpace) ArgMax() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListArgMax{
				ListArgMax: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// Any returns whether any value in each boolean sublist is true.
func (ns ListNameSpace) Any() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListAny{
				ListAny: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// All returns whether all values in each boolean sublist are true.
func (ns ListNameSpace) All() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListAll{
				ListAll: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// DropNulls removes null values from each sublist.
func (ns ListNameSpace) DropNulls() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListDropNulls{
				ListDropNulls: &pb.ListFunction{Expr: ns.expr.toProto()},
			},
		},
	}
}

// CountMatches counts how often the provided value occurs in each sublist.
func (ns ListNameSpace) CountMatches(element interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListCountMatches{
				ListCountMatches: &pb.ListCountMatches{
					Expr:    ns.expr.toProto(),
					Element: exprFromValue(element).toProto(),
				},
			},
		},
	}
}

// Get returns the item at the provided index in each sublist.
func (ns ListNameSpace) Get(index interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListGet{
				ListGet: &pb.ListGet{
					Expr:      ns.expr.toProto(),
					Index:     exprFromValue(index).toProto(),
					NullOnOob: true,
				},
			},
		},
	}
}

// Gather gets items in each sublist by index expression.
func (ns ListNameSpace) Gather(index interface{}, nullOnOob ...bool) Expr {
	nulls := false
	if len(nullOnOob) > 0 {
		nulls = nullOnOob[0]
	}
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListGather{
				ListGather: &pb.ListGather{
					Expr:      ns.expr.toProto(),
					Index:     exprFromValue(index).toProto(),
					NullOnOob: nulls,
				},
			},
		},
	}
}

// GatherEvery gathers every n-th item from each sublist starting at offset.
func (ns ListNameSpace) GatherEvery(n, offset interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListGatherEvery{
				ListGatherEvery: &pb.ListGatherEvery{
					Expr:   ns.expr.toProto(),
					N:      exprFromValue(n).toProto(),
					Offset: exprFromValue(offset).toProto(),
				},
			},
		},
	}
}

// Slice slices each sublist with the provided offset and length.
func (ns ListNameSpace) Slice(offset, length interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListSlice{
				ListSlice: &pb.ListSlice{
					Expr:   ns.expr.toProto(),
					Offset: exprFromValue(offset).toProto(),
					Length: exprFromValue(length).toProto(),
				},
			},
		},
	}
}

// First returns the first item from each sublist.
func (ns ListNameSpace) First() Expr {
	return ns.Get(int64(0))
}

// Last returns the last item from each sublist.
func (ns ListNameSpace) Last() Expr {
	return ns.Get(int64(-1))
}

// Head returns the first n items from each sublist.
func (ns ListNameSpace) Head(n interface{}) Expr {
	return ns.Slice(int64(0), n)
}

// Tail returns the last n items from each sublist.
func (ns ListNameSpace) Tail(n interface{}) Expr {
	count := exprFromValue(n)
	return ns.Slice(Lit(int64(0)).Sub(count.Cast(Int64, false)), count)
}

// Join joins string values in each sublist using the provided separator.
func (ns ListNameSpace) Join(separator interface{}, ignoreNulls ...bool) Expr {
	ignore := false
	if len(ignoreNulls) > 0 {
		ignore = ignoreNulls[0]
	}
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListJoin{
				ListJoin: &pb.ListJoin{
					Expr:        ns.expr.toProto(),
					Separator:   exprFromValue(separator).toProto(),
					IgnoreNulls: ignore,
				},
			},
		},
	}
}

// Shift shifts each sublist by the provided number of periods.
func (ns ListNameSpace) Shift(periods interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListShift{
				ListShift: &pb.ListShift{
					Expr:    ns.expr.toProto(),
					Periods: exprFromValue(periods).toProto(),
				},
			},
		},
	}
}

// Diff calculates the discrete difference inside each sublist.
func (ns ListNameSpace) Diff(periods int64, nullBehavior DiffNullBehavior) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListDiff{
				ListDiff: &pb.ListDiff{
					Expr:         ns.expr.toProto(),
					Periods:      periods,
					NullBehavior: nullBehavior,
				},
			},
		},
	}
}

func (ns ListNameSpace) sample(value interface{}, isFraction bool, opts ...ListSampleOptions) Expr {
	var sampleOpts ListSampleOptions
	if len(opts) > 0 {
		sampleOpts = opts[0]
	}
	seed := uint64(0)
	hasSeed := false
	if sampleOpts.Seed != nil {
		seed = *sampleOpts.Seed
		hasSeed = true
	}
	msg := &pb.ListSample{
		Expr:            ns.expr.toProto(),
		Value:           exprFromValue(value).toProto(),
		WithReplacement: sampleOpts.WithReplacement,
		Shuffle:         sampleOpts.Shuffle,
		Seed:            seed,
		HasSeed:         hasSeed,
	}
	if isFraction {
		return Expr{inner: &pb.Expr{Kind: &pb.Expr_ListSampleFraction{ListSampleFraction: msg}}}
	}
	return Expr{inner: &pb.Expr{Kind: &pb.Expr_ListSampleN{ListSampleN: msg}}}
}

// SampleN samples n values from each sublist.
func (ns ListNameSpace) SampleN(n interface{}, opts ...ListSampleOptions) Expr {
	return ns.sample(n, false, opts...)
}

// SampleFraction samples a fraction of each sublist.
func (ns ListNameSpace) SampleFraction(fraction interface{}, opts ...ListSampleOptions) Expr {
	return ns.sample(fraction, true, opts...)
}

func (ns ListNameSpace) setOperation(other interface{}, op pb.SetOperation) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListUnion{
				ListUnion: &pb.ListSetOperation{
					Expr:  ns.expr.toProto(),
					Other: exprFromValue(other).toProto(),
					Op:    op,
				},
			},
		},
	}
}

// Union returns the set union between list arrays.
func (ns ListNameSpace) Union(other interface{}) Expr {
	return ns.setOperation(other, pb.SetOperation_UNION)
}

// SetDifference returns the set difference between list arrays.
func (ns ListNameSpace) SetDifference(other interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListSetDifference{
				ListSetDifference: &pb.ListSetOperation{
					Expr:  ns.expr.toProto(),
					Other: exprFromValue(other).toProto(),
					Op:    pb.SetOperation_DIFFERENCE,
				},
			},
		},
	}
}

// SetIntersection returns the set intersection between list arrays.
func (ns ListNameSpace) SetIntersection(other interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListSetIntersection{
				ListSetIntersection: &pb.ListSetOperation{
					Expr:  ns.expr.toProto(),
					Other: exprFromValue(other).toProto(),
					Op:    pb.SetOperation_INTERSECTION,
				},
			},
		},
	}
}

// SetSymmetricDifference returns the symmetric difference between list arrays.
func (ns ListNameSpace) SetSymmetricDifference(other interface{}) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_ListSetSymmetricDifference{
				ListSetSymmetricDifference: &pb.ListSetOperation{
					Expr:  ns.expr.toProto(),
					Other: exprFromValue(other).toProto(),
					Op:    pb.SetOperation_SYMMETRIC_DIFFERENCE,
				},
			},
		},
	}
}

File: rust/src/expr_list.rs
use crate::error::BridgeError;
use crate::executor::build_expr;
use crate::proto;
use polars::lazy::dsl::{concat_list, element};
use polars::prelude::*;
use polars::series::ops::NullBehavior;
use polars_ops::prelude::SetOperation;

pub fn build_list_expr(kind: &proto::expr::Kind) -> Option<Result<Expr, BridgeError>> {
    use proto::expr::Kind;

    match kind {
        Kind::Element(_) => Some(Ok(element())),
        Kind::ConcatList(exprs) => Some(build_concat_list(exprs)),
        Kind::ListEval(eval) => Some(build_list_eval(eval)),
        Kind::ListAgg(agg) => Some(build_list_agg(agg)),
        Kind::ListLen(func) => Some(build_unary(func, "ListLen", |expr| expr.list().len())),
        Kind::ListSum(func) => Some(build_unary(func, "ListSum", |expr| expr.list().sum())),
        Kind::ListMean(func) => Some(build_unary(func, "ListMean", |expr| expr.list().mean())),
        Kind::ListMedian(func) => {
            Some(build_unary(func, "ListMedian", |expr| expr.list().median()))
        }
        Kind::ListMax(func) => Some(build_unary(func, "ListMax", |expr| expr.list().max())),
        Kind::ListMin(func) => Some(build_unary(func, "ListMin", |expr| expr.list().min())),
        Kind::ListStd(moment) => Some(build_list_std(moment)),
        Kind::ListVar(moment) => Some(build_list_var(moment)),
        Kind::ListContains(contains) => Some(build_list_contains(contains)),
        Kind::ListSort(sort) => Some(build_list_sort(sort)),
        Kind::ListReverse(func) => Some(build_unary(func, "ListReverse", |expr| {
            expr.list().reverse()
        })),
        Kind::ListUnique(unique) => Some(build_list_unique(unique)),
        Kind::ListNUnique(func) => Some(build_unary(func, "ListNUnique", |expr| {
            expr.list().n_unique()
        })),
        Kind::ListArgMin(func) => Some(build_unary(func, "ListArgMin", |expr| {
            expr.list().arg_min()
        })),
        Kind::ListArgMax(func) => Some(build_unary(func, "ListArgMax", |expr| {
            expr.list().arg_max()
        })),
        Kind::ListAny(func) => Some(build_unary(func, "ListAny", |expr| expr.list().any())),
        Kind::ListAll(func) => Some(build_unary(func, "ListAll", |expr| expr.list().all())),
        Kind::ListDropNulls(func) => Some(build_unary(func, "ListDropNulls", |expr| {
            expr.list().drop_nulls()
        })),
        Kind::ListCountMatches(count) => Some(build_list_count_matches(count)),
        Kind::ListGet(get) => Some(build_list_get(get)),
        Kind::ListGather(gather) => Some(build_list_gather(gather)),
        Kind::ListGatherEvery(gather) => Some(build_list_gather_every(gather)),
        Kind::ListSlice(slice) => Some(build_list_slice(slice)),
        Kind::ListJoin(join) => Some(build_list_join(join)),
        Kind::ListShift(shift) => Some(build_list_shift(shift)),
        Kind::ListDiff(diff) => Some(build_list_diff(diff)),
        Kind::ListSampleN(sample) => Some(build_list_sample(sample, false)),
        Kind::ListSampleFraction(sample) => Some(build_list_sample(sample, true)),
        Kind::ListUnion(set_op) => Some(build_list_set_operation(set_op)),
        Kind::ListSetDifference(set_op) => Some(build_list_set_operation(set_op)),
        Kind::ListSetIntersection(set_op) => Some(build_list_set_operation(set_op)),
        Kind::ListSetSymmetricDifference(set_op) => Some(build_list_set_operation(set_op)),
        Kind::ListUniqueStable(unique) => Some(build_list_unique(unique)),
        _ => None,
    }
}

fn build_concat_list(exprs: &proto::ConcatList) -> Result<Expr, BridgeError> {
    if exprs.exprs.is_empty() {
        return Err(BridgeError::InvalidArgument(
            "ConcatList requires at least one expression".into(),
        ));
    }

    let built = exprs
        .exprs
        .iter()
        .map(build_expr)
        .collect::<Result<Vec<_>, _>>()?;

    concat_list(built).map_err(|e| BridgeError::Execution(format!("ConcatList failed: {}", e)))
}

fn build_list_eval(eval: &proto::ListEval) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&eval.expr, "ListEval")?;
    let evaluation = build_inner_expr(&eval.evaluation, "ListEval")?;
    Ok(expr.list().eval(evaluation))
}

fn build_list_agg(agg: &proto::ListAgg) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&agg.expr, "ListAgg")?;
    let evaluation = build_inner_expr(&agg.evaluation, "ListAgg")?;
    Ok(expr.list().agg(evaluation))
}

fn build_unary<F>(func: &proto::ListFunction, name: &str, op: F) -> Result<Expr, BridgeError>
where
    F: FnOnce(Expr) -> Expr,
{
    let expr = build_inner_expr(&func.expr, name)?;
    Ok(op(expr))
}

fn build_list_contains(contains: &proto::ListContains) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&contains.expr, "ListContains")?;
    let item = build_inner_expr(&contains.item, "ListContains")?;
    Ok(expr.list().contains(item, contains.nulls_equal))
}

fn build_list_sort(sort: &proto::ListSort) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&sort.expr, "ListSort")?;
    let options = SortOptions::default()
        .with_order_descending(sort.descending)
        .with_nulls_last(sort.nulls_last);
    Ok(expr.list().sort(options))
}

fn build_list_get(get: &proto::ListGet) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&get.expr, "ListGet")?;
    let index = build_inner_expr(&get.index, "ListGet")?;
    Ok(expr.list().get(index, get.null_on_oob))
}

fn build_list_gather(gather: &proto::ListGather) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&gather.expr, "ListGather")?;
    let index = build_inner_expr(&gather.index, "ListGather")?;
    Ok(expr.list().gather(index, gather.null_on_oob))
}

fn build_list_gather_every(gather: &proto::ListGatherEvery) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&gather.expr, "ListGatherEvery")?;
    let n = build_inner_expr(&gather.n, "ListGatherEvery")?;
    let offset = build_inner_expr(&gather.offset, "ListGatherEvery")?;
    Ok(expr.list().gather_every(n, offset))
}

fn build_list_slice(slice: &proto::ListSlice) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&slice.expr, "ListSlice")?;
    let offset = build_inner_expr(&slice.offset, "ListSlice")?;
    let length = build_inner_expr(&slice.length, "ListSlice")?;
    Ok(expr.list().slice(offset, length))
}

fn build_list_join(join: &proto::ListJoin) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&join.expr, "ListJoin")?;
    let separator = build_inner_expr(&join.separator, "ListJoin")?;
    Ok(expr.list().join(separator, join.ignore_nulls))
}

fn build_list_count_matches(count: &proto::ListCountMatches) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&count.expr, "ListCountMatches")?;
    let element = build_inner_expr(&count.element, "ListCountMatches")?;
    Ok(expr.list().count_matches(element))
}

fn build_list_shift(shift: &proto::ListShift) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&shift.expr, "ListShift")?;
    let periods = build_inner_expr(&shift.periods, "ListShift")?;
    Ok(expr.list().shift(periods))
}

fn build_list_sample(sample: &proto::ListSample, is_fraction: bool) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(
        &sample.expr,
        if is_fraction {
            "ListSampleFraction"
        } else {
            "ListSampleN"
        },
    )?;
    let value = build_inner_expr(
        &sample.value,
        if is_fraction {
            "ListSampleFraction"
        } else {
            "ListSampleN"
        },
    )?;
    let seed = if sample.has_seed {
        Some(sample.seed)
    } else {
        None
    };
    if is_fraction {
        Ok(expr
            .list()
            .sample_fraction(value, sample.with_replacement, sample.shuffle, seed))
    } else {
        Ok(expr
            .list()
            .sample_n(value, sample.with_replacement, sample.shuffle, seed))
    }
}

fn build_list_diff(diff: &proto::ListDiff) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&diff.expr, "ListDiff")?;
    let null_behavior = match proto::DiffNullBehavior::try_from(diff.null_behavior) {
        Ok(proto::DiffNullBehavior::DiffNullIgnore) => NullBehavior::Ignore,
        Ok(proto::DiffNullBehavior::DiffNullDrop) => NullBehavior::Drop,
        Err(_) => {
            return Err(BridgeError::Unsupported(format!(
                "Unknown ListDiff null behavior: {}",
                diff.null_behavior
            )))
        }
    };
    Ok(expr.list().diff(diff.periods, null_behavior))
}

fn build_list_std(moment: &proto::ListMoment) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&moment.expr, "ListStd")?;
    let ddof = u8::try_from(moment.ddof).map_err(|_| {
        BridgeError::InvalidArgument(format!("ListStd ddof out of range for u8: {}", moment.ddof))
    })?;
    Ok(expr.list().std(ddof))
}

fn build_list_var(moment: &proto::ListMoment) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&moment.expr, "ListVar")?;
    let ddof = u8::try_from(moment.ddof).map_err(|_| {
        BridgeError::InvalidArgument(format!("ListVar ddof out of range for u8: {}", moment.ddof))
    })?;
    Ok(expr.list().var(ddof))
}

fn build_list_unique(unique: &proto::ListUnique) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&unique.expr, "ListUnique")?;
    if unique.maintain_order {
        Ok(expr.list().unique_stable())
    } else {
        Ok(expr.list().unique())
    }
}

fn build_list_set_operation(set_op: &proto::ListSetOperation) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&set_op.expr, "ListSetOperation")?;
    let other = build_inner_expr(&set_op.other, "ListSetOperation")?;
    let op = match proto::SetOperation::try_from(set_op.op) {
        Ok(proto::SetOperation::Union) => SetOperation::Union,
        Ok(proto::SetOperation::Difference) => SetOperation::Difference,
        Ok(proto::SetOperation::Intersection) => SetOperation::Intersection,
        Ok(proto::SetOperation::SymmetricDifference) => SetOperation::SymmetricDifference,
        Err(_) => {
            return Err(BridgeError::Unsupported(format!(
                "Unknown ListSetOperation op: {}",
                set_op.op
            )))
        }
    };
    match op {
        SetOperation::Union => Ok(expr.list().union(other)),
        SetOperation::Difference => Ok(expr.list().set_difference(other)),
        SetOperation::Intersection => Ok(expr.list().set_intersection(other)),
        SetOperation::SymmetricDifference => Ok(expr.list().set_symmetric_difference(other)),
    }
}

fn build_inner_expr(expr: &Option<Box<proto::Expr>>, name: &str) -> Result<Expr, BridgeError> {
    let expr = expr
        .as_ref()
        .ok_or_else(|| BridgeError::PlanSemantic(format!("{name} has no expr")))?;
    build_expr(expr.as_ref())
}
